### PR TITLE
Cknieling/annotated attribute UI

### PR DESCRIPTION
**Pull request by chr11115**
Parse annotations from rule files to create:
- Angle ranges
- Color picker
- Description/Help text
- Directory picker
- File picker
- Enum (dynamic enums not yet supported)
- Percent (will default to 0-1 range instead of 0-100 for now)

Sadly we can't get rid of the slider in float spare parms, since PRM_TYPE_PLAIN only seem to affect non-spare parms in my tests..

### CODE REVIEW

File: src/palladio/SOPAssign.cpp [hunk 1]
```diff
@@ -427,22 +377,51 @@ void SOPAssign::updateAttributes(GU_Detail* detail) {
...
+						case PRM_Type::PRM_FLOAT_RGBA: {
+							const float r = evalFloat(&parm, 0, time);
+							const float g = evalFloat(&parm, 1, time);
+							const float b = evalFloat(&parm, 2, time);
+							const std::wstring colorString = AnnotationParsing::getColorString({r, g, b});
+
+							UT_String stringValue(toOSNarrowFromUTF16(colorString));
+							GA_RWHandleS stringHandle(detail->addStringTuple(attrOwner, attributeName, 1));
+							stringHandle.set(0, stringValue);
+						}
```
Comment (mistafunk) [inline src/palladio/SOPAssign.cpp:423]: missing break and default case handling

File: src/palladio/NodeSpareParameter.cpp [hunk 2]
```diff
@@ -86,42 +87,131 @@ void addParmsFromTemplateArray(OP_Node* node, PRM_Template* spareParmTemplates,
 }
 
 void addParm(OP_Node* node, PRM_Type parmType, const std::wstring& id, const std::wstring& name, PRM_Default defaultVal,
-             PRM_Range* range, const FolderVec& parentFolders) {
+             PRM_Range* range, const FolderVec& parentFolders, const std::wstring& annotation) {
```
Comment (mistafunk) [inline src/palladio/NodeSpareParameter.cpp:90]: consider renaming `annotation` to `description` for clarity

**Review by mistafunk — CHANGES_REQUESTED** [review]
found a couple small issues and there is a compilation issue on MSVC

File: src/palladio/SOPAssign.cpp (hunk 1 above)
Comment (chr11115) [reply src/palladio/SOPAssign.cpp:423]: fixed

File: src/palladio/NodeSpareParameter.cpp (hunk 2 above)
Comment (chr11115) [reply src/palladio/NodeSpareParameter.cpp:90]: done

File: src/palladio/AnnotationParsing.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	std::wstring mValuesAttr;
+	std::vector<std::wstring> mOptions;
+};
+
+using RangeAnnotation = std::pair<double, double>;
+using FileAnnotation = std::vector<std::wstring>;
+using ColorAnnotation = std::array<double, 3>;
+
+struct AttributeAnnotationInfo {
+	const prt::Annotation* mAnnotation;
```
Comment (mistafunk) [inline src/palladio/AnnotationParsing.h:38]: This is dangerous and hard to read in terms of life-time of `mAnnotation`. It points to the RuleInfo object which "lives" in a different translation unit (`SOPAssign::buildUI`). A quick improvement would be to turn this into a const reference via constructor. Also see below.

File: src/palladio/AnnotationParsing.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+std::wstring getColorString(const std::array<float, 3>& rgb);
+
+RangeAnnotation parseRangeAnnotation(const prt::Annotation* mAnnotation);
+
+EnumAnnotation parseEnumAnnotation(const prt::Annotation* mAnnotation);
+
+FileAnnotation parseFileAnnotation(const prt::Annotation* mAnnotation);
+
+AttributeTrait detectAttributeTrait(const prt::Annotation* mAnnotation);
```
Comment (mistafunk) [inline src/palladio/AnnotationParsing.h:53]: Please rename `mAnnotation` to `annotation` and turn them into const references.

**Review by mistafunk — CHANGES_REQUESTED** [review]
Almost there, just a minor issue regarding annotation life-time/pointers...

File: src/palladio/AnnotationParsing.cpp [hunk 5]
```diff
@@ -239,32 +239,47 @@ AttributeTrait detectAttributeTrait(const prt::Annotation* annotation) {
 }
 
 AttributeAnnotationInfo getAttributeAnnotationInfo(const std::wstring& attributeName, const RuleFileInfoUPtr& info) {
-	AttributeAnnotationInfo attributeAnnotationInfo;
+	const prt::Annotation* annotation = nullptr;
+	AttributeTrait attributeTrait = AttributeTrait::NONE;
+	std::wstring description = L"";
```
Comment (mistafunk) [inline src/palladio/AnnotationParsing.cpp:244]: redundant initialization, remove the L"";

File: src/palladio/SOPAssign.cpp [hunk 6]
```diff
@@ -193,36 +104,103 @@ void updateAttributeUIDefaultValues(SOPAssign* node, const std::wstring& style,
...
-					// only support booleans, i.e. don't store folders
-					if (currParmType.getOrdinalType() != PRM_Type::PRM_OrdinalType::PRM_ORD_TOGGLE)
-						continue;
-
-					const int intValue = static_cast<int>(std::get<bool>(it->second));
-
-					node->setInt(attributeName, 0, time, intValue);
-					parm.overwriteDefaults(time);
-					parm.getTemplatePtr()->setFactoryDefaults(parm.getTemplatePtr()->getDefault(0));
+					switch (currParmType.getOrdinalType()) {
```
Comment (mistafunk) [inline src/palladio/SOPAssign.cpp:107]: GCC warning: not all enum items handled in switch... consider adding a default case

**Review by mistafunk — CHANGES_REQUESTED** [review]
Two small issues... feel free to merge after fixing them.

File: src/palladio/AnnotationParsing.cpp (hunk 5 above)
Comment (chr11115) [reply src/palladio/AnnotationParsing.cpp:244]: done

File: src/palladio/SOPAssign.cpp (hunk 6 above)
Comment (chr11115) [reply src/palladio/SOPAssign.cpp:107]: done

**Pull request merged**